Extract `scan_waku_fleet.py` to a separate container, run as part of waku fleet compose

# Description

Currently we run `tests-functional/scripts/scan_waku_fleet.py` script as part of the Docker image entrypoint:

https://github.com/status-im/status-go/blob/c575471b4c502f4e2d999cb3b3890e1004a1babb/_assets/scripts/entrypoint.sh#L3-L8

This is quite a confusing implementation and: 
- unnecessarily pollutes the Dockerfile
- unnecessarily runs the script in **each** new container

# Proposal

The fleet scanning should be extracted into a separate docker image to be executed as part of `docker-compose.waku.yml`. Then scanning result should then be copied as file from the container (or even be fetched with some API).

@at0m1x19 can you please take care of this one next? 
This is becoming a big pain point, recently tests became very flaky both locally and on CI 🙁 

```nim
self = <test_save_or_update_keycard.TestSaveOrUpdateKeycard object at 0x7fc57cc979d0>
backend_new_profile = <function backend_new_profile.<locals>.factory at 0x7fc57c5e0360>
    @pytest.fixture(autouse=True)
    def setup_backends(self, backend_new_profile):
>       self.account = backend_new_profile("sender")
tests-functional/tests/test_save_or_update_keycard.py:13: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests-functional/tests/conftest.py:97: in factory
    backend.init_status_backend()
tests-functional/clients/status_backend.py:155: in init_status_backend
    return self.api_request_json(method, data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <clients.status_backend.StatusBackend object at 0x7fc57c844490>
method = 'InitializeApplication'
data = {'apiLoggingEnabled': True, 'dataDir': '/usr/status-user', 'logEnabled': True, 'logLevel': 'DEBUG', ...}
kwargs = {}, response = <Response [200]>
json_response = {'error': 'failed to open fleets json file: open /usr/status-user/wakufleetconfig.json: no such file or directory'}
err = 'failed to open fleets json file: open /usr/status-user/wakufleetconfig.json: no such file or directory'
    def api_request_json(self, method, data, **kwargs):
        response = self.api_request(method, data, **kwargs)
        try:
            json_response = response.json()
        except JSONDecodeError:
            raise ApiDecodeError("Invalid JSON in response", method=method, payload=response.content)
    
        err = json_response.get("error")
        if err:
>           raise ApiResponseError(str(err), method=method, payload=json_response)
E           clients.api.ApiResponseError: failed to open fleets json file: open /usr/status-user/wakufleetconfig.json: no such file or directory
tests-functional/clients/api.py:68: ApiResponseError
```

--- a/Dockerfile
+++ b/Dockerfile
@@ -49,7 +49,6 @@ RUN mkdir -p /static/configs
 
 COPY --from=builder /go/src/github.com/status-im/status-go/build/bin/status-backend /usr/local/bin/
 COPY --from=builder /go/src/github.com/status-im/status-go/build/bin/push-notification-server /usr/local/bin/
-COPY --from=builder /go/src/github.com/status-im/status-go/tests-functional/scripts/scan_waku_fleet.py /usr/local/bin
 COPY --from=builder /go/src/github.com/status-im/status-go/static/keys/* /static/keys/
 COPY --from=builder /go/src/github.com/status-im/status-go/tests-functional/waku_configs/* /static/configs/
 COPY --from=builder /go/src/github.com/status-im/nim-sds/build/libsds.so /usr/local/lib/

--- a/Dockerfile
+++ b/Dockerfile
@@ -55,10 +55,6 @@ COPY --from=builder /go/src/github.com/status-im/nim-sds/build/libsds.so /usr/lo
 
 ENV LD_LIBRARY_PATH=/usr/local/lib/
 
-COPY _assets/scripts/entrypoint.sh /usr/local/bin/entrypoint.sh
-RUN chmod +x /usr/local/bin/entrypoint.sh
-
-# 30304 is used for Discovery v5
 EXPOSE 8080 8545 30303 30303/udp 30304/udp
-ENTRYPOINT ["/usr/local/bin/entrypoint.sh"]
-CMD ["--help"]
+
+CMD ["status-backend", "--help"]

--- a/_assets/scripts/entrypoint.sh
+++ b/_assets/scripts/entrypoint.sh
@@ -1,4 +1,0 @@
-#!/bin/sh
-
-# Simple entrypoint: execute the CMD from the Dockerfile.
-exec "$@"

--- a/_assets/scripts/entrypoint.sh
+++ b/_assets/scripts/entrypoint.sh
@@ -1,8 +1,4 @@
 #!/bin/sh
 
-# Define $SCAN_WAKU_FLEET to run a command before running the actual app (CMD).
-# This is expected to be used in functional tests to run `scan_waku_fleet.py` script before starting the app.
-$SCAN_WAKU_FLEET
-
-# This will exec the CMD from your Dockerfile, i.e. "npm start"
+# Simple entrypoint: execute the CMD from the Dockerfile.
 exec "$@"

--- a/_assets/scripts/run_functional_tests.sh
+++ b/_assets/scripts/run_functional_tests.sh
@@ -55,6 +55,28 @@ fi
 echo -e "${GRN}Running status-go external dependencies${RST}"
 docker compose -p ${project_name} ${all_compose_files} up -d --build --remove-orphans
 
+# Wait for wakufleet-scanner to finish before running tests. If it fails,
+# there's no point in starting tests because wakufleetconfig.json will not
+# be available for status-backend.
+echo -e "${GRN}Waiting for wakufleet-scanner to complete${RST}"
+scanner_container_id=$(docker compose -p ${project_name} ${all_compose_files} ps -q wakufleet-scanner || true)
+
+if [[ -z "${scanner_container_id}" ]]; then
+  echo -e "${RED}wakufleet-scanner service container not found. Make sure docker-compose.waku.yml defines it correctly.${RST}"
+  exit 1
+fi
+
+scanner_exit_code=$(docker wait "${scanner_container_id}")
+
+if [[ "${scanner_exit_code}" -ne 0 ]]; then
+  echo -e "${RED}wakufleet-scanner failed with exit code ${scanner_exit_code}. See logs below:${RST}"
+  docker logs "${scanner_container_id}" || true
+  echo -e "${RED}Aborting functional tests because wakufleetconfig.json was not generated successfully.${RST}"
+  exit 1
+fi
+
+echo -e "${GRN}wakufleet-scanner completed successfully${RST}"
+
 # Set up virtual environment
 venv_path="${root_path}/.venv"
 

--- a/tests-functional/Dockerfile.waku-fleet-scanner
+++ b/tests-functional/Dockerfile.waku-fleet-scanner
@@ -1,0 +1,7 @@
+FROM python:3.12-slim
+
+WORKDIR /app
+
+COPY scripts/scan_waku_fleet.py /app/scan_waku_fleet.py
+
+ENTRYPOINT ["python", "/app/scan_waku_fleet.py"]

--- a/tests-functional/clients/statusgo_container.py
+++ b/tests-functional/clients/statusgo_container.py
@@ -68,7 +68,7 @@ class StatusGoContainer:
                 },
                 # Named volume created by docker-compose.waku.yml for wakufleetconfig.json
                 f"{docker_project_name}_wakufleetconfig": {
-                    "bind": "/usr/status-user",
+                    "bind": "/usr/status-user/config",
                     "mode": "rw",
                 },
             },

--- a/tests-functional/clients/statusgo_container.py
+++ b/tests-functional/clients/statusgo_container.py
@@ -60,13 +60,17 @@ class StatusGoContainer:
             "labels": {"com.docker.compose.project": docker_project_name},
             "environment": {
                 "GOCOVERDIR": "/coverage/binary",
-                "SCAN_WAKU_FLEET": self.get_waku_fleet_scan_command(),
             },
             "volumes": {
                 coverage_path: {
                     "bind": "/coverage/binary",
                     "mode": "rw",
-                }
+                },
+                # Named volume created by docker-compose.waku.yml for wakufleetconfig.json
+                f"{docker_project_name}_wakufleetconfig": {
+                    "bind": "/usr/status-user",
+                    "mode": "rw",
+                },
             },
             "extra_hosts": {
                 "host.docker.internal": "host-gateway",
@@ -91,24 +95,6 @@ class StatusGoContainer:
         StatusGoContainer.all_containers.append(self)
 
         logging.debug(f"Container {self.container.name} created. ID = {self.container.id}")
-
-    def get_waku_fleet_scan_command(self):
-        """Returns the command string for scanning Waku fleet and generating config"""
-
-        # Known node names from docker compose
-        bootstrap_nodes = "boot-1"
-        static_nodes = "boot-1"  # Add bootnode, otherwise metadata exchange doesn't happen, and Waku light mode doesn't work
-        store_nodes = "store"
-
-        return (
-            "python3 /usr/local/bin/scan_waku_fleet.py "
-            f"--fleet-name {Config.waku_fleet} "
-            f"--cluster-id 16 "  # Cluster ID matches docker-compose.waku.yml
-            f"--bootstrap-nodes {bootstrap_nodes} "
-            f"--store-nodes {store_nodes} "
-            f"--static-nodes {static_nodes} "
-            f"--output {Config.waku_fleets_config}"
-        )
 
     def __del__(self):
         self.stop()

--- a/tests-functional/conftest.py
+++ b/tests-functional/conftest.py
@@ -61,7 +61,7 @@ def pytest_addoption(parser):
         "--waku-fleets-config",
         action="store",
         help="Path to a local JSON file with Waku fleets configuration. Default value is a path to config in Docker to run 2 local waku nodes",
-        default="/usr/status-user/wakufleetconfig.json",
+        default="/usr/status-user/config/wakufleetconfig.json",
     )
     parser.addoption(
         "--waku-fleet",

--- a/tests-functional/docker-compose.waku.yml
+++ b/tests-functional/docker-compose.waku.yml
@@ -60,3 +60,24 @@ services:
     healthcheck:
       test: ["CMD-SHELL", "curl -X GET -H 'Content-Type: application/json' http://0.0.0.0:8645 || exit 0"]
 
+  wakufleet-scanner:
+    build:
+      context: .
+      dockerfile: Dockerfile.waku-fleet-scanner
+    depends_on:
+      - boot-1
+      - store
+    volumes:
+      - wakufleetconfig:/fleet-config
+    command: [
+      "--fleet-name", "status-go.test",
+      "--cluster-id", "16",
+      "--bootstrap-nodes", "boot-1",
+      "--store-nodes", "store",
+      "--static-nodes", "boot-1",
+      "--output", "/fleet-config/wakufleetconfig.json"
+    ]
+
+volumes:
+  wakufleetconfig:
+

--- a/tests-functional/scripts/scan_waku_fleet.py
+++ b/tests-functional/scripts/scan_waku_fleet.py
@@ -5,6 +5,7 @@ import json
 import sys
 import os
 import logging
+import time
 
 from urllib import request
 from typing import Dict, List, Optional, Tuple
@@ -46,10 +47,25 @@ def parse_args():
 def _fetch_debug_info(host: str) -> Optional[Dict]:
     url = f"http://{host}:8645/debug/v1/info"
     logging.info(f"Fetching debug info from Waku node {url}")
-    with request.urlopen(url, timeout=5.0) as resp:
-        assert resp.status == 200
-        data = resp.read()
-        return json.loads(data.decode("utf-8"))
+
+    attempts = 15
+    delay_seconds = 1.0
+
+    for attempt in range(1, attempts + 1):
+        try:
+            logging.info(f"Attempt {attempt}/{attempts} to fetch debug info from {url}")
+            with request.urlopen(url, timeout=5.0) as resp:
+                if resp.status != 200:
+                    logging.error(f"Unexpected status {resp.status} from {url}")
+                    continue
+                data = resp.read()
+                return json.loads(data.decode("utf-8"))
+        except Exception as e:  # noqa: BLE001
+            logging.error(f"Error fetching debug info from {url}: {e}")
+            if attempt < attempts:
+                time.sleep(delay_seconds)
+
+    return None
 
 
 def _first_dns_addr(listen_addrs: List[str]) -> Optional[str]:

